### ISSUE TITLE
Missed TP in DRB 88 and 89

### ISSUE THREAD

**Issue by BradSwain** [coderrect-inc/OpenRace#196]
**Describe the bug**
There are real races in dataracebench cases 88 and 89 where a global object is allocated, then read/written in parallel, but we fail to report the race.

**Expected behavior**
We should report the race.

**To Reproduce**
Run the tool on dataracebench cases 88 and 89, observe that no race is reported.

**Comment by bz-liu** [comment]
For this specific error, the problem now located at `race::generateFunctionSummary(const llvm::Function &func)`:  
https://github.com/coderrect-inc/OpenRace/blob/efbe2f9c865e9b74af2cdb0635298de5ddf53c9f/src/IR/Builder.cpp#L62

When it analyze function `foo()`:
```cpp
void foo()
{
   (*counter)++;  
}
```

The `llvm::dyn_cast` loses the `%ptr` in `llvm::LoadInstruction` and `llvm::StoreInstruction` (corresponds to the load and store of `counter` in `foo`) when input `func` is from openmp fork and `%ptr` is a global var, e.g.,

```cpp
      %0 = load i32*, i32** @counter, align 8, !dbg !21, !tbaa !22 //from DRB088-dynamic-storage-orig-yes.ll, define dso_local void @foo()
      // becomes
      %0 = load i32, i32* null, align 4, !dbg !21, !tbaa !22 //from print out in generateFunctionSummary()
```

the load pointer `@counter` is missing (become `null`), so `pts(%0)` is empty and all the checking afterwards are skipped. 

THEN, I compared the IR instructions (of type `llvm::Instruction`) generated by openrace/pta and the ones in `tests/data/integration/dataracebench/XXX.ll`, most IR instructions are different. 

THEN, I re-compiled ` DRB088-dynamic-storage-orig-yes.c` using `clang -S -emit-llvm -g` to get a new .ll file (on server), and the corresponding IR instruction is:
```cpp
      %1 = load i32*, i32** @counter, align 8, !dbg !17
```
and there are a lot of differences between the new .ll file and the one stored in `/openrace/tests`, e.g., `@__kmpc_fork_call` is missing.

@BradSwain Did I compile it in a wrong way? Any idea why those IR instructions are different?

**Comment by BradSwain** [comment]
The ll files under `tests/data` are compiled with these flags:
```
-g -O1 -mllvm -disable-llvm-optzns -fno-discard-value-names -fopenmp -S -emit-llvm
```

The IR also goes through pre-processing at the start of race detection.
To see the IR that the analysis sees (after pro-processing) you can add `--dump-ir=file.ll` to openrace to make it print the processed IR to a file named `file.ll`

I'm not sure why the load from the global becomes a load from null though...

**Comment by bz-liu** [comment]
There are two errors in pointer analysis code:

(1) The pro-processing removes `@counter` so that we have `null` for global var. Because the constant propagation process determines that the global var `counter` has not been overwritten (so that `C == ""` and has been constant propagated to others):

https://github.com/coderrect-inc/OpenRace/blob/3404dcb7346affd53bae8ce29cdf33b97b80fc5a/src/PreProcessing/Passes/OMPConstantPropPass.cpp#L72-L87

Since the function `hasGlobalOverwritten` is too naive:

https://github.com/coderrect-inc/OpenRace/blob/3404dcb7346affd53bae8ce29cdf33b97b80fc5a/src/PreProcessing/Passes/OMPConstantPropPass.cpp#L37-L44

Only `isa<StoreInst>` is not enough to guarantee `GV` has no overwritten, e.g., `@counter` in DRB088-dynamic-storage-orig-yes.ll, where `counter` is not referenced within a construct and its points-to objects with dynamic storage duration should be shared.
- The source code: 
```cpp 
counter++
```
- The corresponding IR: there is always a `StoreInst` after `LoadInst`:
```cpp
 %0 = load i32*, i32** @counter, align 8, !dbg !21, !tbaa !22  //--> load to %0
 %1 = load i32, i32* %0, align 4, !dbg !26, !tbaa !27
 %inc = add nsw i32 %1, 1, !dbg !26
 store i32 %inc, i32* %0, align 4, !dbg !26, !tbaa !27               //--> store to %0: changed the value of counter
```

 To precisely determine whether `GV` (`counter` in this case) can be overwritten afterwards is complex and expensive,
 which requires inter-procedural data-flow analysis/forward slicing.

*SOLUTION* 
To maintain the performance (especially we will target at large programs later), we skip the expensive analysis. Instead, besides the check of `isa<StoreInst>`, we also check whether there is other load/store use for the users of `GV`; if so, we assume GV might be changed later. 

Or for my opinion, we can remove `hasGlobalOverwritten` because a global var is either a constant or will be updated globally. Meanwhile, clang compiler can do constant propagation, why don't we turn on this optimization and use its result if this has to be done? 

(2) This is quite a big issue: the nodes created by constraint graph (a.k.a., points-to/pointer assignment graph) for pointer and pointer's pointer are not distinguished. Still in DRB088-dynamic-storage-orig-yes.ll:

The related IR is:
```cpp
@counter = common dso_local global i32* null, align 8, !dbg !0

define dso_local void @foo() #0 !dbg !18 {
entry:
  %0 = load i32*, i32** @counter, align 8, !dbg !21, !tbaa !22
  %1 = load i32, i32* %0, align 4, !dbg !26, !tbaa !27
  %inc = add nsw i32 %1, 1, !dbg !26
  store i32 %inc, i32* %0, align 4, !dbg !26, !tbaa !27
  ret void, !dbg !29
}
```

The correct procedure is (I use the node ID in constraint graph below to represent the created node, you can reproduce the node ID by turning on `-dump-pts` when running openrace): 

1. create an object and pointer for the global var `counter`:
```
6 -> obj //object node for counter
8 -> *counter  //ptr node for *counter
pts(8) = {6} //points-to relation
```

2. create pointers and points-to constraints for nodes in `foo()`:
`I32**` is `**counter`, so firstly we need to have a node representing the pointer's pointer of `counter`: 

``` ? -> **counter ``` and ```pts(?) = {8}``` (from addr-of constraint: ? = &8)

I use `?` because this node is missing in the current constraint graph construction and the final result. This is the **root** problem.

Then, create the load edge:
```
  %0 = load i32*, i32** @counter // (src node ID: ?) -load-> (dst node id: 87), where 87 is %0 in one openmp forked thread
  ... 
```
Then, create the copy edge from the load edge:
```
(src node ID: 8) -copy-> (dst node id: 87)
```
and solve it we have:
```
pts(87) = {6}
```

For the other openmp forked thread, it is the same but the dst node is becomes 88: 
```
pts(88) = {6}
```
Finally, when we do the shared memory computation, we have 87 and 88 access the shared memory obj 6.

*SOLUTION*
Distinguish the node for pointer and pointer's pointer, add the addr-of constraints between them.

**Comment by BradSwain** [comment]
closed by #210
